keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,655 @@
+
+## Design Details
+
+### Job API
+
+We extend the Job API to set different policies by which a Job can be declared as succeeded.
+
+```golang
+type JobSpec struct {
+	...
+	// Specifies the policy when the job can be declared as succeeded.
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// These policies are evaluated in order.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +optional
+	// +listType=atomic
+	SuccessPolicy []SuccessPolicy
+}
+
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some indexes.
+type SuccessPolicy struct {
+	// Represents the requirement on the succeeded pods.
+	// Each target must have at least one acceptance criterion of
+	// "succeededIndexes" or "count".
+	//
+	// +optional
+	Target *SuccessPolicyTarget
+}
+
+// SuccessPolicyTarget describes a Job can be succeeded based on succeeded indexes.
+type SuccessPolicyTarget struct {
+	// Specifies a set of required indexes when .spec.completionMode =
+	// "Indexed" in a text format.
+	// The job is declared as succeeded if a set of indexes are succeeded. 
+	// The list of indexes must come within 0 to `.spec.completions` and
+	// must not contain duplicates. At least one element is required. 
+	// The indexes are represented as decimal integers
+	// separated by commas. The numbers are listed in increasing order. Three or
+	// more consecutive numbers are compressed and represented by the first and
+	// last element of the series, separated by a hyphen.
+	// For example, if the completed indexes are 1, 3, 4, 5 and 7, they are
+	// represented as "1,3-5,7".
+	// 
+	// +optional
+	SucceededIndexes *string
+
+	// Specifies the required number of indexes when .spec.completionMode = "Indexed".
+	// Value can be an absolute number (ex: 5) when the job can be declared as succeeded.
+	//
+	// +optional
+	SucceededCount *int
+
+	// Specifies the action to be taken on the not finished (succeeded or failed) pods 
+	// when the job achieved the requirements.
+	// Possible values are:
+	//
+	// - Continue: indicates that all pods wouldn't be terminated.
+	// - Terminate: indicates that not finished pods would be terminated.
+	//
+	// Default value is Terminate.
+	WhenTargetAchieved WhenTargetAchievedSuccessPolicy
+}
+
+// WhenTargetAchievedSuccessPolicy specifies the action to be taken on the pods
+// when the job achieved the requirements.
+// +enum
+type WhenTargetAchievedSuccessPolicy string
+
+const (
+	// All pods wouldn't be terminated when the job reached successPolicy.
+	ContinueWhenTargetAchievedSuccessPolicy WhenTargetAchievedSuccessPolicy = "Continue"
+
+	// Not finished pods would be terminated when the job reached successPolicy.
+	TerminateWhenTargetAchievedSuccessPolicy WhenTargetAchievedSuccessPolicy = "Terminate"
```

Comment:
I don't have strong objections since I'm not sure how we should be verbose.
So, if @deads2k is fine, I'm ok with a suggested API.

Reply:
Finding the line between terseness and clarity isn't science, but I prefer the shorter `Terminate` rather than longer phrases.  I would add qualifying phrases in cases where we need to differentiate two types of Terminate, but I don't think that's the case here.

Reply:
The shorter phrase means longer docs, because we have to make sure to teach the point clearly enough that very few people misunderstand it.

This looks confusing enough that I'd expect to see issues raised against the docs by people who didn't get it right first time. Even it most people get it, we have a lot of end users and a percentage might not.

Reply:
I'm with @sftim here. Maybe something shorter like: `TerminateRemainingPods`.

Not sure what is a good alternative to `Continue`.

Reply:
@deads2k @alculquicondor @sftim Can we discuss the API name in the beta graduation stage? Because we dropped `whenCriterionAchieved` API from the alpha stage.